chore(create-bot): add envalid support for cleaner .env handling

rainrisa opened the pull request:
https://github.com/af/envalid for cleaner .env handling

this pull request will added a new option (default to `true`) for whether to use Envalid or not

```console
? Select features: (Press <space> to select, <a> to toggle all, <i> to invert selection, and <enter> to proceed)
 ◉  🧼 Use Envalid
 ◉  🥰 Setup ESLint with @antfu/eslint-config
 ◯  🌐 Internationalization
 ◉  📨 Event dispatcher
❯◉  ✨ Use TypeScript
 ◉  📦 Initialize git repository
 ◉  🐳 Generate Dockerfile
```

the reason i chose the soap emoji (🧼) is because i previously used the gear emoji (⚙️), but the spacing was not work,
and soap represents "cleaning," which fits the purpose of Envalid

Preview of generated typescript code:

src/env.ts
```ts
import process from 'node:process'

import { cleanEnv, num, str } from 'envalid'
import 'dotenv/config'

const env = cleanEnv(process.env, {
    API_ID: num(),
    API_HASH: str(),
    BOT_TOKEN: str({ default: undefined }),
})

export default env
```

and after that, you can just import it anywhere on your project
```ts
import env from './env.js'
```

this is my first pull request, so i apologize if there is any mistakes, also please feel free to give any feedback or suggestions

teidesu commented:
first of all thanks for the pr!!

while i do think we should promote some kind of env validation (and i actually wanted to add something similar myself, but didnt have the time), im not sure if i like envalid

i usually just use whatever object validation library i have lying around (usually zod) and do something like

```ts
import { z } from 'zod'

export const env = z.object({
  API_ID: z.coerce.number(),
  API_HASH: z.string(),
  // ...
}).parse(process.env)
```

but at the same time i don't really like zod either because it's error messages are a bit ass (esp. in the context of env validation)

rainrisa commented:
thanks for the feedback!

i personally like the Envalid error messages, because they usually print something simple like this:

```console
================================
 Invalid environment variables:
    API_ID: Invalid number input: "abcdf"
 Missing environment variables:
    API_HASH: undefined
================================
```

it’s specifically built for `.env` validation, and it’s pretty nice to read in the terminal
but yeah, i get your concern, especially the proxy thing (can be weird in debugging), and the cjs part too (not really nice in esm-based project)

i also noticed you mentioned Zod's error messages, i think sometimes they’re kinda hard to read)

i’m open to switching to Zod if that helps, or trying something more flexible like a pluggable validator (though i’m not 100% sure how to structure that yet). if that’s something you think fits better with the project, i’d love to try

also, if you think it's better to wait and revisit `.env` validation later, i totally understand

just let me know what direction you’d prefer

rainrisa commented:
sorry for the delay, i’ve updated the PR to replace Envalid with an optional Zod-based .env validation, based on your suggestion

this will added a new option (default to `true`) for whether to use Zod or not

```console
? Select features: (Press <space> to select, <a> to toggle all, <i> to invert selection, and <enter> to proceed)
 ◉  🥰 Setup ESLint with @antfu/eslint-config
❯◉  🧪 Use Zod for env validation
 ◯  🌐 Internationalization
 ◉  📨 Event dispatcher
 ◉  ✨ Use TypeScript
 ◉  📦 Initialize git repository
 ◉  🐳 Generate Dockerfile
```

#### Some changes i think are worth mentioning:

When generating a project, the user is asked for botToken:
  
```console
? Found existing config, do you want to reuse it? Yes
? Bot token (optional): 
```

if botToken is provided, the generated Zod schema makes BOT_TOKEN required:

```ts
const r = z.object({
    API_ID: z.coerce.number(),
    API_HASH: z.string(),
    BOT_TOKEN: z.string(), // required
}).safeParse(process.env)
```
happy to change this if you prefer optional

#### Deno support

i initially tried using `npm:zod` as suggested by [the docs](https://zod.dev/?id=installation), but ran into an error like `jsr.io/zod` not found.
i switched to `@zod/zod`, which worked when i tested, [source](https://github.com/colinhacks/zod/blob/499df780b10dfcce29afba3e811757f7eade9570/packages/zod/jsr.json#L2)

this is my first time working with Deno, so any feedback on the imports or setup would be great!

#### Preview of generated Deno Typescript code:

src/env.ts:

```ts
import '@std/dotenv/load'
import { z } from '@zod/zod'

const r = z.object({
    API_ID: z.coerce.number(),
    API_HASH: z.string(),
    BOT_TOKEN: z.string(),
}).safeParse(Deno.env.toObject())

if (!r.success) {
    throw new Error('Invalid env:\n' + z.prettifyError(r.error))
}

export const env = r.data
```

#### Preview of generated Node Typescript code:

src/env.ts:

```ts
import { z } from 'zod'

const r = z.object({
    API_ID: z.coerce.number(),
    API_HASH: z.string(),
    BOT_TOKEN: z.string(),
}).safeParse(process.env)

if (!r.success) {
    throw new Error('Invalid env:\n' + z.prettifyError(r.error))
}

export const env = r.data
```

let me know if this aligns with your expectations

teidesu commented:
yeah looks nice thanks!
